### PR TITLE
Support BatchToSpaceND with dynamic crops

### PR DESCRIPTION
**Pull request by JiayingGaoo**
Making BatchToSpaceND support dynamic crops

### CODE REVIEW

File: tests/test_backend.py [hunk 1]
```diff
@@ -2443,6 +2443,17 @@ def test_batch_to_spacend(self):
         _ = tf.batch_to_space_nd(input_x, block_size, crop, name=_TFOUTPUT)
         self._run_test_case([_OUTPUT], {_INPUT: input_val})
 
+    @check_opset_min_version(10, "Slice in opset 10 can accept dynamic 'starts' and 'ends'")
+    def test_batch_to_spacend_with_dynamic_crop(self):
+        block_size = [2, 2]
+        crops_val = np.array([[0, 1], [2, 1]], dtype=np.int32)
```
Comment (zhijxu-MS) [inline tests/test_backend.py]: add test to cover crop's end/right is zero or not zero

File: tests/test_backend.py [hunk 2]
```diff
@@ -2443,6 +2443,17 @@ def test_batch_to_spacend(self):
...
+        input_val = np.random.random_sample([40, 3, 5, 100]).astype(np.float32)
+
+        input_x = tf.placeholder(dtype=tf.float32, shape=input_val.shape, name=_TFINPUT)
+        crops = tf.placeholder(dtype=tf.int32, shape=[2, 2], name=_TFINPUT1)
```
Comment (zhijxu-MS) [inline tests/test_backend.py]: add test to cover int64

File: tf2onnx/onnx_opset/tensor.py [hunk 3]
```diff
@@ -1097,36 +1098,86 @@ def version_1(cls, ctx, node, **kwargs):
...
+                ctx.make_node("Squeeze", [slice1], {"axes": [3]},
+                              outputs=node.output, name=node.name, dtypes=dtypes, shapes=shapes)
+            else:
+                kwargs = {**inputs_map, "outputs": node.output}
+                ctx.remove_node(node.name)
+                GraphBuilder(ctx).make_slice(kwargs, name=node.name, dtypes=dtypes, shapes=shapes)
         else:
+            crops_type = ctx.get_dtype(node.input[2])
+            crops_np_type = utils.map_onnx_to_numpy_type(crops_type)
+
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: remove useless empty line

File: tf2onnx/onnx_opset/tensor.py [hunk 4]
```diff
@@ -1097,36 +1098,86 @@ def version_1(cls, ctx, node, **kwargs):
...
+            # Make the node crops being transposed and respectively obtained its two rows by split op
+            crops_trans = ctx.make_node("Transpose", node.inputs[2].output, {"perm": [1, 0]})
+            trans_split = ctx.make_node("Split", [crops_trans.output[0]], output_count=2, attr={"axis": 0})
+
+            # Create the starts node
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: looks like this comment is useless

File: tests/test_backend.py (hunk 1 above)
Comment (JiayingGaoo) [reply tests/test_backend.py]: Thanks for your comment. I have add two tests for the cases ends and right are zero

File: tests/test_backend.py (hunk 2 above)
Comment (JiayingGaoo) [reply tests/test_backend.py]: Thanks! I have added a test to cover int64 type for the crops

File: tf2onnx/onnx_opset/tensor.py (hunk 3 above)
Comment (JiayingGaoo) [reply tf2onnx/onnx_opset/tensor.py]: I have deleted the useless lines and comments.

File: tf2onnx/onnx_opset/tensor.py (hunk 4 above)
Comment (JiayingGaoo) [reply tf2onnx/onnx_opset/tensor.py]: I have deleted the useless lines and comments.

File: tests/test_backend.py [hunk 5]
```diff
@@ -2443,6 +2443,51 @@ def test_batch_to_spacend(self):
         _ = tf.batch_to_space_nd(input_x, block_size, crop, name=_TFOUTPUT)
         self._run_test_case([_OUTPUT], {_INPUT: input_val})
 
+    @check_opset_min_version(10, "Slice in opset 10 can accept dynamic 'starts' and 'ends'")
+    def test_batch_to_spacend_with_dynamic_crop_top_is_zero(self):
```
Comment (zhijxu-MS) [inline tests/test_backend.py]: could you merge these tests into one, like the function "test_conv2d_8"?
and please remember to add comments, for example "[0, 1], [2, 1]" is used to test the case where crop top is zero since converter needs take special action when top is zero.

File: tf2onnx/onnx_opset/tensor.py [hunk 6]
```diff
@@ -1097,36 +1098,76 @@ def version_1(cls, ctx, node, **kwargs):
...
         shapes = node.output_shapes
 
-        if len(input_shape) == 3:
-            # add a squeeze op to convert output into 3d
-            kwargs = {**inputs_map}
-            ctx.remove_node(node.name)
-            slice1 = GraphBuilder(ctx).make_slice(kwargs)
-            ctx.make_node("Squeeze", [slice1], {"axes": [3]},
-                          outputs=node.output, name=node.name, dtypes=dtypes, shapes=shapes)
+        if ctx.get_node_by_output(node.input[2]).is_const():
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: use node.inputs[2] instead of ctx.get_node_by_output(node.input[2])

File: tf2onnx/onnx_opset/tensor.py [hunk 7]
```diff
@@ -1097,36 +1098,76 @@ def version_1(cls, ctx, node, **kwargs):
...
+                              outputs=node.output, name=node.name, dtypes=dtypes, shapes=shapes)
+            else:
+                kwargs = {**inputs_map, "outputs": node.output}
+                ctx.remove_node(node.name)
+                GraphBuilder(ctx).make_slice(kwargs, name=node.name, dtypes=dtypes, shapes=shapes)
         else:
+            crops_type = ctx.get_dtype(node.input[2])
+            crops_np_type = utils.map_onnx_to_numpy_type(crops_type)
+            # Make the node crops being transposed and respectively obtained its two rows by split op
+            crops_trans = ctx.make_node("Transpose", node.inputs[2].output, {"perm": [1, 0]})
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: use node.input[2] instead of node.inputs[2].output

File: tf2onnx/onnx_opset/tensor.py [hunk 8]
```diff
@@ -1097,36 +1098,76 @@ def version_1(cls, ctx, node, **kwargs):
...
+            # Create the original node for ends
+            ends_ori = ctx.make_node("Squeeze", [trans_split.output[1]], attr={"axes": [0]})
+            # create zero const node
+            zero_value = np.array([0]).astype(crops_np_type)
+            zero_const = ctx.make_const(utils.make_name("Const"), zero_value)
+            # Obtain a node by comparing the ends_ori with zero
+            ends_ori_equal_zero = ctx.make_node("Equal", [ends_ori.output[0], zero_const.output[0]])
+            ends_ori_equal_zero = ctx.make_node("Cast", [ends_ori_equal_zero.output[0]], attr={"to": crops_type})
+            # Create int max const node
+            int_max_value = np.array([utils.get_max_value(crops_np_type)]).astype(crops_np_type)
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: do we need astype(crops_np_type) here? I think utils.get_max_value(crops_np_type) will return the correct dtype

File: tf2onnx/onnx_opset/tensor.py [hunk 9]
```diff
@@ -1097,36 +1098,76 @@ def version_1(cls, ctx, node, **kwargs):
...
+            crops_type = ctx.get_dtype(node.input[2])
+            crops_np_type = utils.map_onnx_to_numpy_type(crops_type)
+            # Make the node crops being transposed and respectively obtained its two rows by split op
+            crops_trans = ctx.make_node("Transpose", node.inputs[2].output, {"perm": [1, 0]})
+            trans_split = ctx.make_node("Split", [crops_trans.output[0]], output_count=2, attr={"axis": 0})
+            starts = ctx.make_node("Squeeze", [trans_split.output[0]], attr={"axes": [0]})
+            # Create the original node for ends
+            ends_ori = ctx.make_node("Squeeze", [trans_split.output[1]], attr={"axes": [0]})
+            # create zero const node
+            zero_value = np.array([0]).astype(crops_np_type)
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: could give a detail procedures here as a comment. like https://github.com/onnx/tensorflow-onnx/blob/4d0fcedd2d3331069aaa2c69e3c353e8a8c2b8c1/tf2onnx/onnx_opset/nn.py#L724

so not need to give a comment every step and we can know the whole steps at a glance.

File: tf2onnx/onnx_opset/tensor.py [hunk 10]
```diff
@@ -1097,36 +1098,76 @@ def version_1(cls, ctx, node, **kwargs):
...
+
+            if len(input_shape) == 3:
+                # add a squeeze op to convert output into 3d
+                kwargs = {**inputs_map}
+                ctx.remove_node(node.name)
+                slice1 = GraphBuilder(ctx).make_slice(kwargs)
+                ctx.make_node("Squeeze", [slice1], {"axes": [3]},
+                              outputs=node.output, name=node.name, dtypes=dtypes, shapes=shapes)
+            else:
+                kwargs = {**inputs_map, "outputs": node.output}
```
Comment (zhijxu-MS) [inline tf2onnx/onnx_opset/tensor.py]: since two branch both need this step, could you move it out so that we only call make_slice once.

File: tf2onnx/onnx_opset/tensor.py (hunk 7 above)
Comment (JiayingGaoo) [reply tf2onnx/onnx_opset/tensor.py]: Thanks for your comments.

File: tf2onnx/onnx_opset/tensor.py (hunk 8 above)
Comment (JiayingGaoo) [reply tf2onnx/onnx_opset/tensor.py]: Thanks. I thinks it is needed. Without astype(crops_np_type), there will be an error about unmatched type(int32 and int64).

**Comment by guschmue** [comment]
closing since CLA is not signed.

**Comment by vadimkantorov** [comment]
too bad this did not go in :(

**Comment by guschmue** [comment]
We'd love to take it but we can't take it for legal reasons if the CLA is not signed.
